File: PDOO/Napakalaki-Ruby/lib/Napakalaki.rb
#encoding: UTF-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame
  
  require 'singleton'
  require_relative "monster.rb"
  require_relative "card_dealer.rb"
  require_relative "player.rb"
  require_relative "combat_result.rb"

  class Napakalaki
    include Singleton

    def initialize
      @currentPlayer = nil
      @currentPlayerIndex = 0;
      @players = Array.new
      @dealer = CardDealer.getInstance
      @currentMonster = nil
    end

    def developCombat
      combatResult = @currentPlayer.combat(@currentMonster)
      @dealer.giveMonsterBack(@currentMonster)
      if(combatResult == NapakalakiGame::CombatResult::LOSEANDCONVERT)
        nC = @dealer.nextCultist
        cP = CultistPlayer.new(@currentPlayer, nC)
        @players[@currentPlayerIndex]=cP
        @players.each { |p|  
          if(p.enemy == cP)
            p.setEnemy(cP)
          end
        }
        @currentPlayer = cP
      end
      return combatResult    
    end
  
    def discardVisibleTreasures(treasures)
      treasures.each { |treasure|  
        @currentPlayer.discardVisibleTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      }
    end
  
    def discardHiddenTreasures(treasures)
      treasures.each { |treasure|  
        @currentPlayer.discardHiddenTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      }
    end
  
    def makeTreasuresVisible(treasures)
      treasures.each { |t|
        @currentPlayer.makeTreasureVisible(t)
      }
    end
  
    def initGame(players)
      initPlayers(players)
      setEnemies
      @dealer.initCards
      self.nextTurn
    end
  
    def getCurrentPlayer
      return @currentPlayer
    end
  
    def getCurrentMonster
      return @currentMonster
    end
  
    def nextTurn
      stateOK = nextTurnAllowed
      if (stateOK)
        @currentMonster = @dealer.nextMonster
        @currentPlayer = nextPlayer
        dead = @currentPlayer.isDead
        if(dead)
          @currentPlayer.initTreasures
        end
      end
      return stateOK
    end
  
    def endOfGame(result)
      resultado = true if (result == CombatResult::WINGAME)
      return resultado
    end
  
    private
      
    def initPlayers(names)
      names.each { |i|
        player = Player.new(i)
        @players << player
      }
    end
  
    def nextPlayer
      if(@currentPlayer == nil && @currentPlayerIndex == 0)
        numero_aleatorio=rand(@players.size)
        @currentPlayerIndex = numero_aleatorio
        jugador = @players.at(numero_aleatorio)
      else
        jugador = @players.at((@currentPlayerIndex + 1) % @players.size)
        @currentPlayerIndex = (@currentPlayerIndex + 1) % @players.size
      end
      return jugador
    end
  
    def nextTurnAllowed
      if (@currentPlayer == nil)
        allowed = true
      else
        allowed = @currentPlayer.validState
      end
      return allowed
    end
  
    def setEnemies
      @players.each { |i|  
        numero_aleatorio=rand(@players.size)
        enemy = @players.at(numero_aleatorio)
        if(enemy == i)
          enemy = @players.at((numero_aleatorio + 1) % @players.size)
        end
        i.setEnemy(enemy)
      }
    end
  end

end

File: PDOO/Napakalaki/src/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;

/**
 *
 * @author nestor
 */
public class Napakalaki {

    private static Napakalaki instance = null;
    private Monster currentMonster;
    private CardDealer dealer;
    private ArrayList<Player> players;
    private Player currentPlayer;
    private int currentPlayerIndex;

    private Random r;

    private Napakalaki() {
        currentMonster = null;
        dealer = CardDealer.getInstance();
        players = new ArrayList();
        currentPlayer = null;
        currentPlayerIndex = 0;

        r = new Random();
    }

    private void initPlayers(ArrayList<String> names) {
        for (String name : names) {
            Player player = new Player(name);
            players.add(player);
        }
    }

    private void setEnemies() {
        Player enemigo;
        for (Player player : players) {
            int numero_aleatorio = r.nextInt(players.size());
            enemigo = players.get(numero_aleatorio);

            if (player == enemigo) {
                enemigo = players.get((numero_aleatorio + 1) % players.size());
            }
            player.setEnemy(enemigo);
        }
    }

    public static Napakalaki getInstance() {
        if (instance == null) {
            instance = new Napakalaki();
        }
        return instance;
    }

    public Player nextPlayer() {
        Player jugador;
        if (currentPlayer == null && currentPlayerIndex == 0) {
            int numero_aleatorio = r.nextInt(players.size());
            currentPlayerIndex = numero_aleatorio;
            jugador = players.get(numero_aleatorio);
        } else {
            jugador = players.get((currentPlayerIndex + 1) % players.size());
            currentPlayerIndex = (currentPlayerIndex + 1) % players.size();
        }

        return jugador;
    }

    public boolean nextTurnAllowed() {
        boolean allowed;
        if (currentPlayer == null) {
            allowed = true;
        } else {
            allowed = currentPlayer.validState();
        }
        return allowed;
    }

    public CombatResult developCombat() {
        CombatResult combatResult = currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        if (combatResult == CombatResult.LOSEANDCONVERT) {
            Cultist nC = dealer.nextCultist();
            CultistPlayer cP = new CultistPlayer(currentPlayer, nC);
            players.set(currentPlayerIndex, cP);
            for(Player p : players){
                if(p.enemy == currentPlayer)
                    p.setEnemy(cP);
            }
            currentPlayer = cP;           
        }
        return combatResult;
    }

    public void discardVisibleTreasures(ArrayList<Treasure> treasures) {
        Iterator<Treasure> it = treasures.iterator();
        while (it.hasNext()) {
            Treasure treasure = it.next();
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }

    public void discardHiddenTreasures(ArrayList<Treasure> treasures) {
        Iterator<Treasure> it = treasures.iterator();
        while (it.hasNext()) {
            Treasure treasure = it.next();
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }

    public void makeTreasuresVisible(ArrayList<Treasure> treasures) {
        Iterator<Treasure> it = treasures.iterator();
        while (it.hasNext()) {
            Treasure t = it.next();
            currentPlayer.makeTreasureVisible(t);
        }
    }

    public void initGame(ArrayList<String> players) {
        this.initPlayers(players);
        this.setEnemies();
        dealer.initCards();
        this.nextTurn();
    }

    public Player getCurrentPlayer() {
        return currentPlayer;
    }

    public Monster getCurrentMonster() {
        return currentMonster;
    }

    public boolean nextTurn() {
        boolean stateOK = this.nextTurnAllowed();
        if (stateOK) {
            currentMonster = dealer.nextMonster();
            currentPlayer = this.nextPlayer();
            boolean dead = currentPlayer.isDead();
            if (dead) {
                currentPlayer.initTreasures();
            }
        }
        return stateOK;
    }

    public boolean endOfGame(CombatResult result) {
        boolean resultado;
        if (result == CombatResult.WINGAME) {
            resultado = true;
        } else {
            resultado = false;
        }
        return resultado;
    }
}
